karabo_bridge/client.py

```diff
@@ -66,10 +69,15 @@ def __init__(self, endpoint, sock='REQ', ser='msgpack'):
         else:
             raise NotImplementedError('socket is not supported:', str(sock))
 
+        if timeout is not None:
+            self._socket.setsockopt(zmq.RCVTIMEO, timeout)
+        self._recv_ready = False
```

Comment:
I'm a bit wary of this extra state variable. I haven't thought through in detail, but I think it might end up with both server and client thinking it's their turn to listen for a message, and it won't be obvious why nothing is getting sent.

I don't have a better idea in the short term, but maybe this is the real reason to use another socket type like PUSH/PULL - it avoids having client & server state which can get out of step.

Reply:
I also don't like it so much... But is has proven reliable at least in use with karaboFAI.
I should spend more time playing with PUSH/PULL and other patterns, but (PUSH/PULL) seems to bring also some more complication:
- handling properly message queues, since messages are very large it can easily crash in client applications
- PUSH/PULL doesn't allow load balancing

Reply:
I think we can constrain the memory usage easily enough in our client code by setting a HWM. Not sure what you mean about the load balancing. I'm thinking about how the bridge can work most efficiently; maybe this is something we can work on when I'm there next week.

Reply:
The PUSH socket does not allow load balancing if you have several clients (PULL sockets), it is shared in round robin manner, so if a client is slow or not ready it will still receive data that will be queued. It is generally bad if you have some client that is used for monitoring once in a while what is in the data, in that case it will consume 1/nth train even if not using anything - also thinking what happens if one client is not consuming data, and the input queue is full, it this blocking the whole interface?)... Also for tools like onda, I believe the master process is waiting for processed train/pulses to arrive in order so if a slow worker still queue data at its input it can increase delay quite a bit.

Reply:
And I never managed to make use of the HWM, if you have an example if it working I'm highly interested!

Reply:
The HWM definitely has an effect. I've put a couple of scripts [in a gist](https://gist.github.com/takluyver/a916abfe368010333c345b1fcb3d47c7) - if you want to experiment, I've been running one `push.py` and a couple of `pull.py`.

I can't completely make sense of the numbers with smaller messages, but when I send 4MB messages, I can get the queue waiting for each pull socket down to 3 messages, which I think is one queued on the receiver, one queued on the sender, and one in the process of being transferred. A HWM of 0 is not an option (setting it to zero means unlimited).

If we do want to go down that route, I think it would be necessary to split up trains so that each pulse is sent as a separate message, so we've got more, smaller messages to work with. I think this has other benefits as well: if it can take up to a second to transfer data for one train, it's really useful if processing can start on the first pulse while later ones are still being sent.

Reply:
Thanks,

I tried it quickly, it does seems to work! I however hit into something weird:
It happened that exiting a PULL socket while receiving data could crash the PUSH socket:

```bash
sent 177
Assertion failed: !more (src/lb.cpp:110)
zsh: abort (core dumped)  python pushpull.py push
```

Bug is known and fixed in `libzmq 4.2.x`
https://github.com/zeromq/libzmq/issues/1588

I had to update `pyzmq>=17.1.0` to be built against the lib 4.2, but then the weird behavior is: when closing the PULL socket while receiving data the PUSH is not crashing anymore, but the next data received (after restarting a PULL) starts from the last part not sent. :

```bash
 % python pushpull.py pull
RCVHWM 1
RCVBUF -1
# parts: 2
# parts: 2
# parts: 2
^C%
tmichela@machine ~/projects/tmp/pushpull
 % python pushpull.py pull
RCVHWM 1
RCVBUF -1
# parts: 1
# parts: 2
# parts: 2
# parts: 2
```

you can find the code for the example [here](https://gist.github.com/tmichela/ef979ede7efadc7bcc4a75041af8c5cf)

Reply:
it also seem that closing a PULL socket affect other PULL sockets:

```bash
 % python pushpull.py pull
RCVHWM 1
RCVBUF -1
9 - # parts: 2
10 - # parts: 3
11 - # parts: 3
12 - # parts: 3
13 - # parts: 3
14 - # parts: 3
16 - # parts: 3
19 - # parts: 3
20 - # parts: 3
23 - # parts: 3
25 - # parts: 3
26 - # parts: 2
27 - # parts: 3
29 - # parts: 3
31 - # parts: 2
32 - # parts: 3
33 - # parts: 3
34 - # parts: 3
36 - # parts: 3
38 - # parts: 2
39 - # parts: 3
```

Here when I receive partial messages when an other PULL socket is terminated.

Reply:
Shall we move this discussion to a separate issue? I think it's possible that's a bug with ZeroMQ or pyzmq.

Reply:
Opened an issue on pyzmq for it: https://github.com/zeromq/pyzmq/issues/1244

Reply:
Thanks!